### PR TITLE
Enhance AbstractDummy4jBuilder, refactor and format code and javadocs…

### PR DESCRIPTION
**Pull request by little-pinecone**
…, refactor empty checks.

Add AbstractDummy4jBuilder::locale and AbstractDummy4jBuilder::paths methods that accept varrargs. Add new tests for AbstractDummy4jBuilder. Refactor checking for null or empty locale and paths, ensure falling back to default values when necessary. Change MissingLocaleException class name to MissingLocaleDefinitionException and update its message.

### CODE REVIEW

File: src/main/java/dev/codesoapbox/dummy4j/AbstractDummy4jBuilder.java [hunk 1]
```diff
@@ -34,5 +36,33 @@ public T paths(List<String> paths) {
...
+        if (theOnlyElementIsNull(locale)) {
+            self().locale = new ArrayList<>();
+        } else {
+            self().locale = Arrays.asList(locale);
+        }
+        return self();
+    }
+
+    private boolean theOnlyElementIsNull(String[] elements) {
+        return elements.length == 1 && elements[0] == null;
```
Comment (daniel-frak) [inline src/main/java/dev/codesoapbox/dummy4j/AbstractDummy4jBuilder.java]: Not sure what to rename it but you could encapsulate even more logic in the method:
```java
private List<String> varArgsToList() {
  if(elements.length == 1 && elements[0] == null) {
    return emptyList();
  }
  return Arrays.asList(elements);
}
```
Then:
```java
public T locale(String... locale) {
  self().locale = varArgsToList(locale);
  return self();
}
```

File: src/test/java/dev/codesoapbox/dummy4j/Dummy4jBuilderTest.java [hunk 2]
```diff
@@ -19,26 +23,155 @@ void shouldBuildWithSeed() {
...
-        // Standard path should build fine
-        new Dummy4jBuilder().build();
+    void shouldBuildWithDefaultSeedLocaleAndPaths() {
+        Dummy4j dummy4j = new Dummy4jBuilder().build();
+
+        assertNotNull(dummy4j);
+    }
+
+    @Test
+    void shouldBuildWithGivenAndValidLocaleAndPaths() {
```
Comment (daniel-frak) [inline src/test/java/dev/codesoapbox/dummy4j/Dummy4jBuilderTest.java]: Valid=Existing?

File: src/test/java/dev/codesoapbox/dummy4j/Dummy4jBuilderTest.java [hunk 3]
```diff
@@ -19,26 +23,155 @@ void shouldBuildWithSeed() {
...
+                .locale(locale)
+                .build();
+
+        assertNotNull(dummy4j);
+    }
+
+    @Test
+    void shouldFallBackToDefaultLocaleAndPathsWhenEmptyArrayIsGiven() {
+        Dummy4j bothEmpty = new Dummy4jBuilder()
+                .paths(new ArrayList<>())
```
Comment (daniel-frak) [inline src/test/java/dev/codesoapbox/dummy4j/Dummy4jBuilderTest.java]: Perhaps `emptyList()` will communicate intent better here?

File: src/test/java/dev/codesoapbox/dummy4j/Dummy4jBuilderTest.java [hunk 4]
```diff
@@ -19,26 +23,155 @@ void shouldBuildWithSeed() {
...
+        Dummy4j dummy4j = new Dummy4jBuilder()
+                .paths(path)
+                .locale(locale)
+                .build();
+
+        assertNotNull(dummy4j);
+    }
+
+    @Test
+    void shouldFallBackToDefaultLocaleAndPathsWhenEmptyArrayIsGiven() {
```
Comment (daniel-frak) [inline src/test/java/dev/codesoapbox/dummy4j/Dummy4jBuilderTest.java]: `WhenEmptyListsAreGiven`

File: src/test/java/dev/codesoapbox/dummy4j/Dummy4jBuilderTest.java [hunk 5]
```diff
@@ -19,26 +23,155 @@ void shouldBuildWithSeed() {
...
+    }
+
+    @ParameterizedTest
+    @CsvSource({
+            "nonExistingPath, en",
+            "dummy4j, nonExistingLocale",
+            "nonExistingPath, nonExistingLocale"
+    })
+    void shouldThrowExceptionWhenPathOrLocaleDontExist(String path, String locale) {
+        Dummy4jBuilder fromArrays = new Dummy4jBuilder()
```
Comment (daniel-frak) [inline src/test/java/dev/codesoapbox/dummy4j/Dummy4jBuilderTest.java]: `fromLists`?

File: src/test/java/dev/codesoapbox/dummy4j/Dummy4jBuilderTest.java [hunk 6]
```diff
@@ -19,26 +23,155 @@ void shouldBuildWithSeed() {
...
+                .paths(singletonList(path))
+                .locale(singletonList(locale));
+
+        Dummy4jBuilder fromStrings = new Dummy4jBuilder()
```
Comment (daniel-frak) [inline src/test/java/dev/codesoapbox/dummy4j/Dummy4jBuilderTest.java]: `fromVarArgs`?

**Review by daniel-frak — CHANGES_REQUESTED** [review]

File: src/main/java/dev/codesoapbox/dummy4j/AbstractDummy4jBuilder.java [hunk 7]
```diff
@@ -46,16 +46,12 @@ public T locale(String... locale) {
     }
 
     private List<String> varArgsToList(String... elements) {
-        if (theOnlyElementIsNull(elements)) {
+        if (elements.length == 1 && elements[0] == null) {
```
Comment (little-pinecone) [inline src/main/java/dev/codesoapbox/dummy4j/AbstractDummy4jBuilder.java:44]: What if every element in varrargs is null?

Comment (daniel-frak) [reply src/main/java/dev/codesoapbox/dummy4j/AbstractDummy4jBuilder.java:44]: Same as with `Arrays.asList(null, null, null);`, so probably not our problem.

File: src/main/java/dev/codesoapbox/dummy4j/AbstractDummy4jBuilder.java [hunk 8]
```diff
@@ -34,5 +37,28 @@ public T paths(List<String> paths) {
         return self();
     }
 
+    /**
+     * @since SNAPSHOT
+     */
+    public T locale(String... locale) {
```
Comment (daniel-frak) [inline src/main/java/dev/codesoapbox/dummy4j/AbstractDummy4jBuilder.java:38]: Group methods by name.

File: src/test/java/dev/codesoapbox/dummy4j/Dummy4jBuilderTest.java [hunk 9]
```diff
@@ -19,26 +23,155 @@ void shouldBuildWithSeed() {
...
+        );
+    }
+
+    @ParameterizedTest
+    @CsvSource({
+            "nonExistingPath, en",
+            "dummy4j, nonExistingLocale",
+            "nonExistingPath, nonExistingLocale"
+    })
+    void shouldThrowExceptionWhenPathOrLocaleDontExist(String path, String locale) {
```
Comment (daniel-frak) [inline src/test/java/dev/codesoapbox/dummy4j/Dummy4jBuilderTest.java]: `Dont`->`DoesNot`

**Review by daniel-frak — CHANGES_REQUESTED** [review]

**Pull request merged**